UltraStar Play/Assets/Scenes/Sing/PlayerController/PlayerNoteRecorder.cs

```diff
@@ -90,125 +87,159 @@ public List<RecordedNote> GetRecordedNotes(Sentence sentence)
     }
 
-    private void HandleEndOfLastRecordedNote(double currentBeat)
+    private void HandleRecordedNoteStarted(int midiNote, double currentBeat)
     {
-        // End the note seamlessly, i.e., continue the last recorded note until now.
-        lastRecordedNote.EndBeat = currentBeat;
-        LimitRecordedNoteBoundsToTargetNoteBounds(lastRecordedNote);
-        playerController.OnRecordedNoteEnded(lastRecordedNote);
-        lastRecordedNote = null;
-    }
+        Sentence currentSentence = playerController.CurrentSentence;
 
-    private void HandleContinuedNote(double currentBeat)
-    {
-        if (lastRecordedNote == null)
+        // Only accept recorded notes where a note is expected in the song
+        Note noteAtCurrentBeat = GetNoteAtBeat(currentSentence, currentBeat);
+        if (noteAtCurrentBeat == null)
         {
             return;
         }
 
-        if (playerController == null)
+        // If the last note ended at the start of the new note, then continue using the last ended note.
+        int roundedMidiNote = GetRoundedMidiNoteForRecordedNote(noteAtCurrentBeat, midiNote);
+        double startBeat = Math.Floor(currentBeat);
+        if (lastEndedNote != null
+            && lastEndedNote.Sentence == currentSentence
+            && lastEndedNote.EndBeat == startBeat
+            && lastEndedNote.RoundedMidiNote == roundedMidiNote)
         {
-            lastRecordedNote = null;
+            lastRecordedNote = lastEndedNote;
+            HandleRecordedNoteContinued(currentBeat);
             return;
         }
-        Sentence currentSentence = playerController.CurrentSentence;
-        if (currentSentence == null)
+
+        lastRecordedNote = new RecordedNote(midiNote, Math.Floor(currentBeat), currentBeat);
+        // The note at the same beat is the target note that should be sung
+        lastRecordedNote.TargetNote = noteAtCurrentBeat;
+        lastRecordedNote.Sentence = currentSentence;
+        lastRecordedNote.RoundedMidiNote = roundedMidiNote;
+
+        // Remember this note
+        AddRecordedNote(lastRecordedNote, currentSentence);
+    }
+
+    private void HandleRecordedNoteContinued(double currentBeat)
+    {
+        lastRecordedNote.EndBeat = currentBeat;
+
+        bool targetNoteIsDone = (lastRecordedNote.TargetNote != null && lastRecordedNote.EndBeat >= lastRecordedNote.TargetNote.EndBeat);
+        if (targetNoteIsDone)
         {
-            lastRecordedNote = null;
-            return;
+            lastRecordedNote.EndBeat = lastRecordedNote.TargetNote.EndBeat;
+            playerController.OnRecordedNoteEnded(lastRecordedNote);
+        }
+        else
+        {
+            playerController.OnRecordedNoteContinued(lastRecordedNote);
         }
 
-        // Only accept recorded notes where a note is expected in the song
-        Note noteAtCurrentBeat = GetNoteAtBeat(currentSentence, currentBeat);
-        if (noteAtCurrentBeat == null)
+        if (targetNoteIsDone)
```

Comment:
please integrate this into the same if check a few lines above

Reply:
done